### PR TITLE
Fix DB serialization order for Ord types

### PR DESCRIPTION
**Pull request by Leo-Besancon**
An error prone behaviour in our code is that when serializing numbers (or slots) for the keys of the DB, we use varint serializer. As a result, the ordering of such numbers is not garantied in the DB's ordering (for instance when the number is around a power of two).

This is e.g. related to the following issue and associated PR: https://github.com/massalabs/massa/issues/4101 and https://github.com/massalabs/massa/pull/4104

In this PR, we will replace every instance of such serialization (and deserialization).

TODO:
- [x] List every DB key and check whether it is impacted 
- [x] Fix serialization of the keys in the DB
- [x] Fix deserialiazation of the keys in the DB (also in the is_db_key_value_valid) check function!)

Potentially impacted files list. For each file, list the potentially impacted keys:
- [x] https://github.com/massalabs/massa/blob/main/massa-async-pool/src/pool.rs @Leo-Besancon
  - recompute_message_info_cache iterates over the DB (https://github.com/massalabs/massa/blob/main/massa-async-pool/src/pool.rs#L228)
  - fetch_message(message_id) iterates over all the keys with the message_id as a prefix.
  - is_key_value_valid(message_id, message) deserializes the message_id key
  - put_entry(&self, message_id: &AsyncMessageId, message: AsyncMessage, batch: &mut DBBatch) serializes the message_id
  - Same for update_entry(&self, message_id: &AsyncMessageId, message_update: AsyncMessageUpdate, batch: &mut DBBatch)
  - Same for delete_entry
  - For a given (async_message_id, async_message), it is serialized in the DB with:
    - ASYNC_POOL_PREFIX + AsyncMessageIdSerializer::serialize(message_id) + MESSAGE_PART_IDENT
- [x] https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs @AurelienFT 
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L174 => We iter but is not a problem if we don’t have the right order
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L181C34-L181C34 => Shouldn't be an issue as we are not caring about the key value.
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L303 => We iter but is not a problem if we don’t have the right order
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L326 =>  Shouldn't be an issue as we are not caring about the key value.
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L371 =>  Shouldn't be an issue as we are not caring about the key value.
  - https://github.com/massalabs/massa/blob/main/massa-db-worker/src/massa_db.rs#L706 => Iter  Shouldn't be an issue as we are not caring about the key value.
- [x] https://github.com/massalabs/massa/blob/main/massa-executed-ops/src/executed_ops.rs @AurelienFT => => No problem for me as keys contains only prefix + operation Id which doesn’t have a variant and are not meant to be sorted
- [x] https://github.com/massalabs/massa/blob/main/massa-executed-ops/src/executed_denunciations.rs @AurelienFT
  - https://github.com/massalabs/massa/blob/main/massa-executed-ops/src/executed_denunciations.rs#L63 => Iter but shouldn’t be a problem because it iter on the whole content of denunciations
  - Serialisation in order is needed broken but not assumed anywhere in this file so it’s not a breaking problem.
- [x] https://github.com/massalabs/massa/blob/main/massa-final-state/src/final_state.rs @Eitu33
- `pos_state.create_new_cycle_from_last`: function from PoS, looks like a potential issue but haven’t looked all the way as I was not assigned to that file
lines are:
L234
L303
L351
L380
- `pos_state.apply_changes_to_batch`: takes slot as input and uses its serialized version when extending deferred credits
side note: other apply_changes functions in finalize taking slot as input are only using it for pruning purposes
lines are:
L450
- when checking db validity, iterates over the whole database and deserializes values to check validity, order should have no impact here
L556-688
- deletes:
reset values using constant prefixes
L128-135
L848-858
delete a cycle info with prefix, prefix is built with cycle number
L224
L280
L398

- [x] https://github.com/massalabs/massa/blob/main/massa-ledger-worker/src/ledger_db.rs @Eitu33 
- key_serializer_db & key_deserializer_db usage:
takes care only of ledger keys, which isnt the problem here (version byte + addr serialize + key type)
side note: to see all of them you can just query “serializer_db” the file
- iterators based on prefixes (again only using ledger key formatting)
L191
L463
L490 (test_export)
L526 (test_export)
- delete_prefix with base prefix input (basically just a clear)
L214

- [x] https://github.com/massalabs/massa/blob/main/massa-module-cache/src/hd_cache.rs @Leo-Besancon
  - Is not impacted (only Hash::to_bytes as keys)
- [x] https://github.com/massalabs/massa/blob/main/massa-pos-exports/src/pos_final_state.rs @Leo-Besancon
  - Every key in CYCLE_HISTORY_PREFIX starts with the cycle_prefix serialized as a Varint.
    - complete_key
    - rng_seed_key
    - final_state_hash_snapshot_key
    - roll_count_prefix
    - roll_count_key
    - prod_stats_prefix
    - upper_limit_prefix
    - prod_stats_fail_key
    - prod_stats_success_key
  - Deferred_credits starts with the slot serialized as a Varint of period and varint of thread
- [x] https://github.com/massalabs/massa/blob/main/massa-versioning/src/versioning.rs @sydhds 
  * Every key have MIP Info serialized with some varint (version & component version)
    * when we init from the db (e.g. from disk) we load everything, store the udpates in a BTreeMap (thus sorting the MIP Info) then call the update_with functions. This can become a problem if we have way too much MIP info stored in db (> 100 000) ? - Should be discussed..

Impacted keys:
- [x] Critical:
  - [x] All the keys that are prefixed by a cycle (pos_final_state.rs)
  - [x] Deferred credits key (contains a slot - pos_final_state.rs)
- [x] Non critical instances that we won't change in this PR (to avoid too many unnecessary changes that could introduce other problems): 
  - [x] In the AsyncPool, the keys in the DB are AsyncMessageId that are Ord and partly encoded as Varint. We do not (currently) use this Ord, but may in the future).
  - [x] Executed Denunciations are indexed by DenunciationIndexed that are Ord and partly encoded as Varint.. We do not (currently) use this Ord, but may in the future)
  - [x] MIP Info related keys, but we don't iter on them in the DB, we query all of them at start. 

# OTHER FACTORS

* [x] make sure network ban times are 1 day
* [x] update version to MAIN.2.0
* [x] update package versions
* [x] add validated Genesis files
  * [x] wait for validation
  * [x] update files
  * [x] update readme

### CODE REVIEW

**Comment by damip** [comment]
For slots that are used in a key: use Slot::to_bytes_key() and Slot::from_bytes_key()

For integers, use to_be_bytes() and from_be_bytes()

**Comment by AurelienFT** [comment]
My part is done.

**Comment by damip** [comment]
My independent analysis (I went through everything except versioning/MIP):

https://github.com/massalabs/massa/pull/4620/files

**Comment by Eitu33** [comment]
final state and ledger db done

**Comment by Leo-Besancon** [comment]
I checked every change made in https://github.com/massalabs/massa/pull/4620/files , this PR seems to take into account the same ser/deser.

**Review by AurelienFT — APPROVED** [review]

File: massa-pos-exports/src/pos_final_state.rs
Comment (damip) [inline massa-pos-exports/src/pos_final_state.rs:733]: A bit more compact and less error prone:

```rust
            // deserialize the slot
            let rest = &serialized_key[DEFERRED_CREDITS_PREFIX.len()..];
            let slot = Slot::from_bytes_key(
                rest.get(..SLOT_KEY_SIZE)
                    .and_then(|v| v.try_into().ok())
                    .expect(DEFERRED_CREDITS_DESER_ERROR),
            );
            let rest = &rest[SLOT_KEY_SIZE..];
```

**Comment by damip** [comment]
All those variables on slices are stressing me out a bit.
Here is a less error-prone approach:

```
fn buf_to_array_ctr<F, V, const N: usize>(
    buffer: &[u8],
    constructor: F,
) -> Result<(&[u8], V), ModelsError>
where
    F: Fn(&[u8; N]) -> V,
{
    if buffer.len() < N {
        return Err(ModelsError::BufferError("buffer too short".into()));
    }
    let v = constructor(
        &buffer[..N].try_into().unwrap(), // checked above
    );
    Ok((&buffer[N..], v))
}
```

Then you can do:

```
let (rest_key, slot) =
    buf_to_array_ctr(rest_key, Slot::from_bytes_key).expect("error !");

let (rest_key, cycle) =
    buf_to_array_ctr(rest_key, |v| u64::from_be_bytes(v.clone()))
        .expect("error !");
```

File: massa-pos-exports/src/pos_final_state.rs
Comment (damip) [inline massa-pos-exports/src/pos_final_state.rs:998]: what is this ?

Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs:998]: This basically reverts the previous fix (that avoided the assumption that cycles are in order in the DB) made in https://github.com/massalabs/massa/pull/4104.

This is not needed, but it simplifies a bit the code.

Comment (damip) [reply massa-pos-exports/src/pos_final_state.rs:998]: do we have tests to ensure that the change won't cause a regression ?

Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs:998]: test_pos_cache_recomputation() at the bottom of the file wil catch this. It was the test I added in #4104 to make sure the cycle_history is correctly recovered from the db (in the correct order)

File: massa-pos-exports/src/pos_final_state.rs
Comment (bilboquet) [inline massa-pos-exports/src/pos_final_state.rs:722]: what about
```rust
        let iterator_cf = db.iterator_cf(
            STATE_CF,
            MassaIteratorMode::From(&start_key_buffer, MassaDirection::Forward),
        ).filter(|(k, _)| k.starts_with(DEFERRED_CREDITS_PREFIX.as_bytes()));

        for (serialized_key, serialized_value) in
        iterator_cf
        {
            // if !serialized_key.starts_with(DEFERRED_CREDITS_PREFIX.as_bytes()) {
            //     // println!("break !start_with");
            //     break;
            // }
```
if I'm correct, it's equivalent and based on few measures the loop is 10% faster

Comment (damip) [reply massa-pos-exports/src/pos_final_state.rs:722]: isn't there a risk that it iterates starting from `start_key_buffer` and until **the end of the whole database** ? I'm afraid `filter()` is not short-circuiting, contrary to `https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while`

Comment (bilboquet) [reply massa-pos-exports/src/pos_final_state.rs:722]: Yes, `take_while` is the correct version of what I wanted to do.
In my test, I counted the number of entries at the end and they were the same. So I was wrong, but the conditions were favorable so the timing looks correct.
I tried again with `take_while` and it seems to be a bit faster (probably because it short-circuits)

Comment (Leo-Besancon) [reply massa-pos-exports/src/pos_final_state.rs]: In another PR, we may want to go over similar behaviour in the code and see if it's worth it to change. (/!\ some of the DB iteration have weird logic, e.g. the iter over async_messages in a comment above, we have to make sure that everything is the same before changing it).

File: massa-execution-worker/src/execution.rs
Comment (bilboquet) [inline massa-execution-worker/src/execution.rs:1845]: I think we could merge those 2 let bindings into one like this:
```rust
        let res_final: BTreeMap<Slot, Amount> = self
            .final_state
            .read()
            .get_pos_state()
            .get_deferred_credits_range(.., Some(address))
            .credits
            .iter()
            .filter_map(|(slot, addr_amount)| {
                addr_amount.get(address).map(|amount| (*slot, *amount))
            })
            .collect();
```

**Review by sydhds — DISMISSED** [review]

**Review by AurelienFT — DISMISSED** [review]

File: massa-execution-worker/src/execution.rs
Comment (Leo-Besancon) [reply massa-execution-worker/src/execution.rs]: Done

**Review by AurelienFT — DISMISSED** [review]

**Review by bilboquet — APPROVED** [review]

**Review by sydhds — DISMISSED** [review]

**Review by damip — DISMISSED** [review]

**Comment by damip** [comment]
Only the genesis files update remains. Waiting for external approval

**Review by damip — DISMISSED** [review]

**Pull request merged**
